torch/distributed/_tensor/experimental/attention.py

```diff
@@ -243,37 +294,59 @@ def _templated_ring_attention(
     key = key.contiguous()
     value = value.contiguous()
 
-    sdpa_merger = _SDPAMerger(_convert_to_f32)
+    sdpa_merger = _SDPAMerger(_cp_options.convert_to_f32, seq_dim=seq_dim)
 
     rest: List[Any]
     out: torch.Tensor
     logsumexp: torch.Tensor
 
     for i in range(size):
-        # overlap communication with compute
         if next_kv is not None:
+            # Wait for the kv from the (cp_rank - 1) rank.
             next_kv = _maybe_wait(next_kv)
             key = next_kv[: key.numel()].reshape(key.shape)
             value = next_kv[key.numel() :].reshape(value.shape)
 
         if i < (size - 1):
+            # Send the k, v to the next rank
             next_kv = torch.cat([key.flatten(), value.flatten()])
             next_kv = _ring_rotate(next_kv, pg, send_to_next=True)
 
         is_causal_behavior = _is_causal_behavior(
             rank=rank, world_size=size, i=i, is_causal=is_causal
         )
 
-        if is_causal_behavior != _CausalBehavior.SKIP:
-            out, logsumexp, *rest = op(
+        if is_causal_behavior == _CausalBehavior.SKIP:
+            continue
+
+        if i == 0 or (not _cp_options.enable_load_balance or not is_causal):
+            # We need to do SDPA with the full local q, k, v.
+            q, k, v, partial = (query, key, value, False)
+        elif i <= rank:
+            # Striped load balancing case, and i <= rank.
+            # We need to do SPDA, with only the first half of the k, v.
+            # Note that q, k, v, each contains two stripes.
+            q, k, v, partial = (
                 query,
-                key,
-                value,
-                is_causal=is_causal_behavior.value,
-                **kwargs,
+                key.chunk(2, dim=2)[0],
+                value.chunk(2, dim=2)[0],
+                False,
             )
-
-            sdpa_merger.step(out, logsumexp)
+        else:
+            # Striped load balancing case, and i > rank.
+            # We need to do SPDA with only the second half of the q, and update
+            # only the the second part of  logsumexp. So partial is True.
+            # Note that q, k, v, each contains two stripes.
+            q, k, v, partial = query.chunk(2, dim=2)[1], key, value, True
+
+        out, logsumexp, *rest = op(
```

Comment:
it is actually not easy to trace through the math, becuase it is not clear what 'op' is doing. do you know if the various SDPA kernels have some docstring somewhere? i did not find documentation for the _ring or _efficient kernels.  (only for the regular SDPA kernel, which does not do the same thing as these other kernels)

Reply:
Add a comment.